### PR TITLE
Bug fix - Short lifecycle objects

### PR DESCRIPTION
**Pull request by BenNaccarato**
#### Description
- Fix the flow for actors that are created, and then cleaned up before spatial has responded to the create entity requets (in this case, we will not have authority, so the delete entity request is ignored, but the channel is cleaned up regardless). For this case, we check if we receive authority over an actor for which we don't have a channel, and assume that this is due to us deleting it before we have authority, and so send a request to delete it then.
- (This check also fixes a crash bug where we attempt to use a channel that is null in UpdateShadowData.)
- Add a check to not create actors and channels for actors that are torn off.
- Also demote some of the regular warnings to verbose.

#### Release note
Bugfix - Avoid crashes and bizarre behaviour when deleting recently created actors.

#### Tests
Tested locally (in and out of PIE) for UT. Larger-scale playtest with QA still to-do

#### Documentation
None necessary.

#### Primary reviewers
@Vatyx @m-samiec

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 1]
```diff
@@ -394,6 +402,14 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
 				return;
 			}
 
+			// If the received actor is torn off, don't bother receiving it.
+			if (EntityActor->GetTearOff())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: stretch goal - can we check tearoff in the component data, and save the unnecessary create/destroy?

Comment (BenNaccarato) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Can look into it 👍

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 2]
```diff
@@ -249,24 +249,32 @@ void USpatialReceiver::HandleActorAuthority(Worker_AuthorityChangeOp& Op)
...
 			{
-				Actor->Role = ROLE_Authority;
-				Actor->RemoteRole = ROLE_SimulatedProxy;
-
-				if (Actor->IsA<APlayerController>())
-				{
-					Actor->RemoteRole = ROLE_AutonomousProxy;
-				}
-				else if (APawn* Pawn = Cast<APawn>(Actor))
+				if (IsValid(NetDriver->GetActorChannelByEntityId(Op.entity_id)))
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Could we use `!= nullptr`?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -327,6 +335,30 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
...
+	// If the received actor is torn off, don't bother receiving it.
+	// Check the pending add components, to find the root component for the received entity.
+	for (PendingAddComponentWrapper& PendingAddComponent : PendingAddComponents)
+	{
+		if (PendingAddComponent.EntityId != EntityId
+			|| ClassInfoManager->GetCategoryByComponentId(PendingAddComponent.ComponentId) != SCHEMA_Data)
+		{
+			continue;
+		}
+		uint32 componentOffset = 0;
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Change to `Offset` to follow naming elsewhere.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -1433,7 +1433,7 @@ void USpatialNetDriver::RemoveActorChannel(Worker_EntityId EntityId)
 {
 	if (!EntityToActorChannel.Contains(EntityId))
 	{
-		UE_LOG(LogSpatialOSNetDriver, Warning, TEXT("RemoveActorChannel: Failed to find entity/channel mapping for entity %lld."), EntityId);
+		UE_LOG(LogSpatialOSNetDriver, Verbose, TEXT("RemoveActorChannel: Failed to find entity/channel mapping for entity %lld."), EntityId);
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: What's the legitimate case this actor shows up? There's UNR-709 you should pick up.

Comment (BenNaccarato) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: So I believe the main reason that this happens is that the channel is cleaned up on tearoff. It is then also (attempted to be) cleaned up when the entity is deleted. It's safe in this case to just try and fail to remove the channel.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 2 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: `IsValid` checks for null as well as PendingKill.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -327,6 +335,30 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
...
+		if (!ClassInfoManager->GetOffsetByComponentId(PendingAddComponent.ComponentId, componentOffset) || componentOffset != 0)
+		{
+			continue;
+		}
+
+		Worker_ComponentData* ComponentData = static_cast<improbable::DynamicComponent*>(PendingAddComponent.Data.Get())->Data;
+		Schema_Object* ComponentObject = Schema_GetComponentDataFields(ComponentData->schema_type);
+		if (Schema_GetBool(ComponentObject, SpatialConstants::ACTOR_TEAROFF_ID))
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Can we add a sanity check in SchemaGenerator that the tearoff property actually matches this ID? (similar to `ACTOR_COMPONENT_REPLICATES_ID`)

Comment (BenNaccarato) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Will look into this 👍

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 6]
```diff
@@ -565,11 +597,11 @@ void USpatialReceiver::DestroyActor(AActor* Actor, Worker_EntityId EntityId)
 	{
 		if (Actor == nullptr)
 		{
-			UE_LOG(LogSpatialReceiver, Warning, TEXT("Removing actor as a result of a remove entity op but cannot find the actor channel! EntityId: %lld"), EntityId);
+			UE_LOG(LogSpatialReceiver, Verbose, TEXT("Removing actor as a result of a remove entity op, which has a missing actor channel. EntityId: %lld"), EntityId);
 		}
 		else
 		{
-			UE_LOG(LogSpatialReceiver, Warning, TEXT("Removing actor as a result of a remove entity op but cannot find the actor channel! Actor: %s EntityId: %lld"), *Actor->GetName(), EntityId);
+			UE_LOG(LogSpatialReceiver, Verbose, TEXT("Removing actor as a result of a remove entity op, which has a missing actor channel. Actor: %s EntityId: %lld"), *Actor->GetName(), EntityId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: I think it's worth keeping this message as a Warning, as I'm not sure when it'd happen in normal flow.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 4 above)
Comment (BenNaccarato) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Looks like it" also outputs during a normal authoritative actor deletion, as we clean up the channel and entity locally when we send the entity deletion request. When the remove op comes back then, the actor and channel is already gone."

**Pull request merged**